# Parsing Libraries Sample Code

In [ ]:
!pip install pdfplumber

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.2/48.2 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.2/60.2 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 72.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 79.4 MB/s eta 0:00:00


In [ ]:
import pdfplumber

with pdfplumber.open("/content/Wang_Marketing_Analytics_Resume.pdf") as pdf, open("output.txt", "w", encoding="utf-8") as f:

    for page in pdf.pages:
        t = page.extract_text()
        if t:
            f.write(t + '\n')


In [12]:
!pip install PyMuPDF

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0/20.0 MB 63.1 MB/s eta 0:00:00


In [ ]:
import fitz  # PyMuPDF

doc = fitz.open("/content/Wang_Marketing_Analytics_Resume.pdf")
with open("output_2.txt", "w", encoding="utf-8") as f:

    for page in doc:
        f.write(page.get_text() + '\n')

In [ ]:
doc = fitz.open("/content/sample_1.pdf")
with open("sample_1_output.txt", "w", encoding="utf-8") as f:

    for page in doc:
        f.write(page.get_text() + '\n')

# Studying Documents To Be Parsed

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import os
def load_files_from_directory(directory_path):
    """
    Loads all files from the specified directory.

    Args:
        directory_path (str): The path to the directory.

    Returns:
        list: A list of file paths.
    """
    file_names = [] # with no file format mark
    years = []
    countries = []
    s = set()
    year_country_pairs = []
    try:
        for filename in os.listdir(directory_path):
            filepath = os.path.join(directory_path, filename)
            if os.path.isfile(filepath):
                file_name = filename
                file_names.append(file_name)
                year = file_name.split("_")[0]
                country = file_name.split("_")[1]
                year_country_pair = year + '_' + country
                years.append(year)
                countries.append(country)
                s.add(filename)
                year_country_pairs.append(year_country_pair)

    except FileNotFoundError:
        print(f"Error: Directory not found: {directory_path}")
        return None
    except Exception as e:
         print(f"An error occurred: {e}")
         return None

    print("Number of files in this folder: ", len(file_names))
    print("Number of unique files in this folder: ", len(s))
    return file_names, len(s), years, countries, year_country_pairs

# Inaug Files

In [47]:
# directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/TextFiles"
# TextFiles_filenames, num_unique_files, _,_,_ = load_files_from_directory(directory_path)
# print("TextFiles number of unique files: ", num_unique_files)
# print(TextFiles_filenames)

directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_pdf"
inaug_pdf_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("inaug_pdf_file number of unique files: ", num_unique_files)
print(inaug_pdf_filenames)

directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt"
processed_inaug_txt_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("processed_inaug_txt_file number of unique files: ", num_unique_files)
print(processed_inaug_txt_filenames)

Number of files in this folder:  150
Number of unique files in this folder:  150
inaug_pdf_file number of unique files:  150
['20141201_Denmark_National Cybersecurity Strategy_DAN.pdf', '20121207_Norway_National Cybersecurity Strategy_NRW.pdf', '20110802_South Korea_National Cybersecurity Strategy_ENG.pdf', '20141102_Uganda_National Cybersecurity Strategy_ENG.pdf', '20130101_Spain_National Cybersecurity Strategy_ESP.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_RUS.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_ENG.pdf', '20160223_Uzbekistan_National Cybersecurity Strategy_RUS.pdf', '20170101_Chile_National Cybersecurity Strategy_ESP.pdf', '20060101_Malaysia_National Cybersecurity Strategy_ENG.pdf', '20160718_Bulgaria_National Cybersecurity Strategy_BGR.pdf', '20170417_Paraguay_National Cybersecurity Strategy_ESP.pdf', '20131201_Portugal_National Cybersecurity Strategy_POR.pdf', '20140101_Latvia_National Cybersecurity Strategy_LTV.pdf', '20180701_Macedonia_Nati

In [46]:
# Double check file numbers
directory_path_pdf = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_pdf"
directory_path_txt = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt"

def compare_folders(folder_a, folder_b):
    # Get file names in each folder
    files_a, files_b = os.listdir(folder_a), os.listdir(folder_b)
    files_a, files_b = set(file[:-4] for file in files_a), set(file[:-4] for file in files_b)

    print(files_a, files_b)
    # Intersection: files in both folders
    common_files = files_a & files_b

    # Files only in folder A
    only_in_a = files_a - files_b

    # Files only in folder B
    only_in_b = files_b - files_a

    return common_files, only_in_a, only_in_b

common_files, only_in_pdf, only_in_txt = compare_folders(directory_path_pdf, directory_path_txt)
print("Number of files in pdf folder: ", len(os.listdir(directory_path_pdf)))
print("Number of files in txt folder: ", len(os.listdir(directory_path_txt)))

print("Number of files in both folders: ", len(common_files))
print("Number of files only in pdf folder: ", len(only_in_pdf))
print("Number of files only in txt folder: ", len(only_in_txt))

print(only_in_pdf)
print(only_in_txt)

{'20180307_Greece_National Cybersecurity Strategy_ENG', '20110101_Luxembourg_National Cybersecurity Strategy_FRN', '20160718_Bulgaria_National Cybersecurity Strategy_ENG', '20141201_Denmark_National Cybersecurity Strategy_DAN', '20180717_North Macedonia_National Cybersecurity Strategy_ENG', '20141102_Uganda_National Cybersecurity Strategy_ENG', '20140101_Bangladesh_National Cybersecurity Strategy_ENG', '20171001_Costa Rica_National Cybersecurity Strategy_ENG', '20150101_Malta_National Cybersecurity Strategy_ENG', '20190101_Seychelles_National Cybersecurity Strategy_ENG', '20101220_Burkina Faso_National Cybersecurity_FRN', '20210309_Vanuatu_National Cybersecurity Strategy_ENG', '20190101_Belarus_National Cybersecurity Strategy_ENG', '20110629_Lithuania_National Cybersecurity Strategy_ENG', '20190601_Lebanon_National Cybersecurity Strategy_ENG', '20170101_Kazakhstan_National Cybersecurity Strategy_KAZ', '20101003_Canada_National Cybersecurity Strategy_ENG', '20160315_Ukraine_National Cyb

In [19]:
import fitz  # PyMuPDF

def process(folder_to_be_processed, index):
  root = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_pdf'
  pdf_file_to_be_processed = folder_to_be_processed[index]
  file_path = "/".join([root, pdf_file_to_be_processed]) + '.pdf'
  doc = fitz.open(file_path)
  txt_folder_path = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt'
  with open(txt_folder_path + '/' + pdf_file_to_be_processed[:-4] + ".txt", "w", encoding="utf-8") as f:
      for page in doc:
          f.write(page.get_text() + '\n')

folder_to_be_processed = ['20121207_Norway_National Cybersecurity Strategy_NRW']
for index in range(len(folder_to_be_processed)):
  process(folder_to_be_processed, index)

## Previous Inaug Parsing

In [ ]:
text_files = set(filename[:-4] for filename in TextFiles_filenames)
inaug_pdf_filenames = set(filename[:-4] for filename in inaug_pdf_filenames)
intersection = list(set(text_files) & set(inaug_pdf_filenames))
print('There are', len(intersection), 'files in inaug pdfs that are already processed and stored in text files')
print(intersection)

There are 68 files in inaug pdfs that are already processed and stored in text files
['20170101_Bahrain_National Cybersecurity Strategy_ENG', '20170321_Malawi_National Cybersecurity Strategy_ENG', '20101003_Canada_National Cybersecurity Strategy_ENG', '20110101_Netherlands_National Cybersecurity Strategy_ENG', '20130101_Spain_National Cybersecurity Strategy_ENG', '20140501_Qatar_National Cybersecurity Strategy_ENG', '20130101_Finland_National Cybersecurity Strategy_ENG', '20170101_Mexico_National Cybersecurity Strategy_ENG', '20151204_South Africa_National Cybersecurity Strategy_ENG', '20090601_United Kingdom_National Cybersecurity Strategy_ENG', '20160101_Samoa_National Cybersecurity Strategy_ENG', '20171023_Armenia_National Cybersecurity Strategy_ENG', '20121207_Norway_National Cybersecurity Strategy_ENG', '20050101_Philippines_National Cybersecurity Strategy_ENG', '20151007_Croatia_National Cybersecurity Strategy_ENG', '20120101_Georgia_National Cybersecurity Strategy_ENG', '2017062

In [ ]:
already_processed_txt = set(filename[:-4] for filename in processed_inaug_txt_filenames)
print(already_processed_txt)

{'20000909_Russia_National Cybersecurity Strategy_ENG', '20030201_USA_National Cybersecurity Strategy_ENG', '20120423_Cyprus_National Cybersecurity Strategy_ENG'}


In [ ]:
to_be_processed = []
for inaug_pdf_filename in inaug_pdf_filenames:
  if inaug_pdf_filename not in intersection:
    if inaug_pdf_filename not in already_processed_txt:
      to_be_processed.append(inaug_pdf_filename + '.pdf')
    else: print("This pdf is already processed and stored in inaug_txt_folder.")
print('There are', len(to_be_processed), 'files in inaug pdfs that needs to be processed.')
print(to_be_processed)

This pdf is already processed and stored in inaug_txt_folder.
This pdf is already processed and stored in inaug_txt_folder.
This pdf is already processed and stored in inaug_txt_folder.
There are 80 files in inaug pdfs that needs to be processed.
['20150409_Peru_National Cybersecurity Strategy_ESP.pdf', '20250201_Liechtenstein_National Cybersecurity Strategy.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_ENG.pdf', '20200101_Tunisia_National Cybersecurity Strategy_FRN.pdf', '20220101_Dominican Republic_National Cybersecurity Strategy_ESP.pdf', '20140101_Syria_National Cybersecurity Strategy_Arabic.pdf', '20161206_Iraq_National Cybersecurity Strategy_IRQ.pdf', '20240101_Togo_National Cybersecurity Strategy_FRN.pdf', '20200101_Eswatini_National Cybersecurity Strategy.pdf', '20170101_Kazakhstan_National Cybersecurity Strategy_KAZ.pdf', '20190101_Seychelles_National Cybersecurity Strategy.pdf', '20121207_Norway_National Cybersecurity Strategy_NRW.pdf', '20200101_Belize_National 

In [ ]:
intersection = [filename + '.txt' for filename in intersection]
print(f"There are {len(intersection)} inaug files in TextFiles.")
print(intersection)

There are 68 inaug files in TextFiles.
['20170101_Bahrain_National Cybersecurity Strategy_ENG.txt', '20170321_Malawi_National Cybersecurity Strategy_ENG.txt', '20101003_Canada_National Cybersecurity Strategy_ENG.txt', '20110101_Netherlands_National Cybersecurity Strategy_ENG.txt', '20130101_Spain_National Cybersecurity Strategy_ENG.txt', '20140501_Qatar_National Cybersecurity Strategy_ENG.txt', '20130101_Finland_National Cybersecurity Strategy_ENG.txt', '20170101_Mexico_National Cybersecurity Strategy_ENG.txt', '20151204_South Africa_National Cybersecurity Strategy_ENG.txt', '20090601_United Kingdom_National Cybersecurity Strategy_ENG.txt', '20160101_Samoa_National Cybersecurity Strategy_ENG.txt', '20171023_Armenia_National Cybersecurity Strategy_ENG.txt', '20121207_Norway_National Cybersecurity Strategy_ENG.txt', '20050101_Philippines_National Cybersecurity Strategy_ENG.txt', '20151007_Croatia_National Cybersecurity Strategy_ENG.txt', '20120101_Georgia_National Cybersecurity Strategy_

In [ ]:
import os
i = 0
for test_file in intersection:
  i += 1
  older_directory = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/TextFiles"
  new_directory = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt"
  os.replace(older_directory + '/' + test_file, new_directory + '/' + test_file)
  print('Moved ', i, 'th file ', test_file, ' in TextFiles to the new folder.')

Moved  1 th file  20170101_Bahrain_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  2 th file  20170321_Malawi_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  3 th file  20101003_Canada_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  4 th file  20110101_Netherlands_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  5 th file  20130101_Spain_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  6 th file  20140501_Qatar_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  7 th file  20130101_Finland_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  8 th file  20170101_Mexico_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  9 th file  20151204_South Africa_National Cybersecurity Strategy_ENG.txt  in TextFiles to the new folder.
Moved  10 th file  2009060

In [ ]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/TextFiles"
TextFiles_filenames, num_unique_files, _,_,_ = load_files_from_directory(directory_path)
print(TextFiles_filenames)

Number of files in this folder:  149
Number of unique files in this folder:  149
['20080101_Oman_ICT Policy_ENG.txt', '20180101_Togo_ICT Policy_ENG.txt', '20131201_Italy_ICT Policy_ENG.txt', '20140101_Qatar_ICT Policy_ENG.txt', '20160610_Japan_ICT Policy_ENG.txt', '20100511_Japan_ICT Policy_ENG.txt', '20120101_India_ICT Policy_ENG.txt', '20060202_Japan_ICT Policy_ENG.txt', '20150101_Qatar_ICT Policy_ENG.txt', '20060101_China_ICT Policy_ENG.txt', '20090203_Japan_ICT Policy_ENG.txt', '20120101_Egypt_ICT Policy_ENG.txt', '20150110_Italy_ICT Policy_ENG.txt', '20140101_Kenya_ICT Policy_ENG.txt', '20140101_Mexico_ICT Policy_ENG.txt', '20120101_Jordan_ICT Policy_ENG.txt', '20130101_Malawi_ICT Policy_ENG.txt', '20150101_Bhutan_ICT Policy_ENG.txt', '20071201_Norway_ICT Policy_ENG.txt', '20140201_Uganda_ICT Policy_ENG.txt', '20171101_Rwanda_ICT Policy_ENG.txt', '20060401_Zambia_ICT Policy_ENG.txt', '20150101_Rwanda_ICT Policy_ENG.txt', '20161101_Jordan_ICT Policy_ENG.txt', '20180101_Gambia_ICT P

In [ ]:
import fitz  # PyMuPDF
def process(index):
  pdfs_folder_name = 'inaugural_cyber_strategies_pdf'
  root = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0'
  file_path = "/".join([root, pdfs_folder_name, to_be_processed[index]])
  doc = fitz.open(file_path)
  folder_path = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt'
  with open(folder_path + '/' + to_be_processed[index][:-4] + ".txt", "w", encoding="utf-8") as f:
      for page in doc:
          f.write(page.get_text() + '\n')

In [ ]:
for index in range(len(to_be_processed)):
  process(index)
  print(index, to_be_processed[index])

0 20150409_Peru_National Cybersecurity Strategy_ESP.pdf
1 20250201_Liechtenstein_National Cybersecurity Strategy.pdf
2 20031105_Tajikistan_National Cybersecurity Strategy_ENG.pdf
3 20200101_Tunisia_National Cybersecurity Strategy_FRN.pdf
4 20220101_Dominican Republic_National Cybersecurity Strategy_ESP.pdf
5 20140101_Syria_National Cybersecurity Strategy_Arabic.pdf
6 20161206_Iraq_National Cybersecurity Strategy_IRQ.pdf
7 20240101_Togo_National Cybersecurity Strategy_FRN.pdf
8 20200101_Eswatini_National Cybersecurity Strategy.pdf
9 20170101_Kazakhstan_National Cybersecurity Strategy_KAZ.pdf
10 20190101_Seychelles_National Cybersecurity Strategy.pdf
11 20121207_Norway_National Cybersecurity Strategy_NRW.pdf
12 20200101_Belize_National Cybersecurity Strategy_ENG.pdf
13 20101220_Burkina Faso_National Cybersecurity Strategy_French.pdf
14 20190101_Belarus_National Cybersecurity Strategy.pdf
15 20220601_Palau_National Cybersecurity Strategy.pdf
16 20200101_Libya_National Cybersecurity Strate

In [ ]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_pdf"
inaug_pdf_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("inaug_pdf_file number of unique files: ", num_unique_files)
print(inaug_pdf_filenames)

directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt"
processed_inaug_txt_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("processed_inaug_txt_file number of unique files: ", num_unique_files)
print(processed_inaug_txt_filenames)

Number of files in this folder:  151
Number of unique files in this folder:  151
inaug_pdf_file number of unique files:  151
['20121123_Belgium_National Cybersecurity Strategy_Dutch.pdf', '20141201_Denmark_National Cybersecurity Strategy_DAN.pdf', '20110101_Luxembourg_National Cybersecurity Strategy_FRH.pdf', '20121207_Norway_National Cybersecurity Strategy_NRW.pdf', '20110802_South Korea_National Cybersecurity Strategy_ENG.pdf', '20141102_Uganda_National Cybersecurity Strategy_ENG.pdf', '20130101_Spain_National Cybersecurity Strategy_ESP.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_RUS.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_ENG.pdf', '20160223_Uzbekistan_National Cybersecurity Strategy_RUS.pdf', '20101220_Burkina Faso_National Cybersecurity Strategy_French.pdf', '20170101_Chile_National Cybersecurity Strategy_ESP.pdf', '20170101_Thailand_National Cybersecurity Strategy_Thai.pdf', '20060101_Malaysia_National Cybersecurity Strategy_ENG.pdf', '20160718_B

# OCR Worklist, Inaug Strategy

Inaug Files Hard to Parse:


**OCR Worklist:**
*  20240101_Togo_National Cybersecurity Strategy_FRN.txt: only parsed page numbers
*  20210309_Vanuatu_National Cybersecurity Strategy.txt: some gibberish at the beginning
*  20210301_Zambia_National Cybersecurity Strategy.txt: some gibberish at the beginning
*   20200101_Benin_National Cybersecurity Strategy_FRN.txt: only address & dates are parsed
*   20200101_Belize_National Cybersecurity Strategy_ENG.txt: weird indenting & hard to read
*   20180701_Macedonia_National Cybersecurity Strategy_ENG.txt: only page number was parsed
*   20151204_South Africa_National Cybersecurity Strategy_ENG.txt: all gibberish
*   20181101_Sri Lanka_National Cybersecurity Strategy_ENG.txt: gibberish
*   20140417_Uruguay_National Cybersecurity Strategy_ESP.txt: empty (equivalent to gibberish)
*   20130312_Panama_National Cybersecurity Strategy_ESP.txt: only page number parsed
*   20140101_Latvia_National Cybersecurity Strategy_ENG.txt: only parsed address
*   20121123_Belgium_National Cybersecurity Strategy_Dutch.txt: empty (equivalent to gibberish)
*   20120831_Chad_National Cybersecurity Strategy.txt: empty (equivalent to gibberish)
*   20110802_South Korea_National Cybersecurity Strategy_ENG.txt: empty (equivalent to gibberish)
*   20180301_Guatemala_National Cybersecurity Strategy_ESP.txt: some pages at the beginning have only page number parsed; there are many grid charts and numbers txt layers, which interfered parsing
*   20170101_Indonesia_National Cybersecurity Strategy_ENG.txt: weird indenting --- original file has complex structure; will try using OCR
*   20140101_Syria_National Cybersecurity Strategy_Arabic.txt: weird indenting --- Arabic & unique formatting; will try using OCR
*   20190601_UAE_National Cybersecurity Strategy.txt: bad indenting --- original file has complex structural graphics, not sure if OCR is going to work on this one

**Progresses until May 19**

*   **OCR failed ones:** 20190601_UAE_National Cybersecurity Strategy(ENG): complex structures & graphics

Note: structurally complex pages like double columns & text-rich-illustrations can lead to random breaks between two paragraphs. OCR parses double column texts by finishing all texts in the left column and then moves on the right column, so some formats under this convention can look strange in txt.

*   **Special Treatment (uploaded):**
  *   20210309_Vanuatu_National Cybersecurity Strategy(ENG): only the table of content is messy, so I replaced gibberish characters with placeholder; the rest of the file is pretty clean
  *   20170101_Indonesia_National Cybersecurity Strategy_ENG: since it only has 8 pages in total, I manually added white boxes to each illustration on each page to process. The output file is pretty clean
  *   20120831_Chad_National Cybersecurity Strategy(FRN; bilingual): this file is actually bilingual. The main text has two columns, one in French and one in Arabic, but the OCR package only supports one language at a time. The parsed txt is in French, and the Arabic part is gibberish -- solution: crop the Arabic section
  *   20181101_Sri Lanka_National Cybersecurity Strategy_ENG: cropped figures
  *   20200101_Belize_National Cybersecurity Strategy_ENG: cropped figures
  *   20180301_Guatemala_National Cybersecurity Strategy_ESP: cropped figures

*   **OCR succeeded ones:**
  *   20210301_Zambia_National Cybersecurity Strategy(uploaded)
  *   20140101_Syria_National Cybersecurity Strategy_Arabic (uploaded)
  *   20200101_Benin_National Cybersecurity Strategy_FRN(uploaded)
  *   20121123_Belgium_National Cybersecurity Strategy_FRN(uploaded)
  *   20140417_Uruguay_National Cybersecurity Strategy_ESP(uploaded)
  *   20110802_South Korea_National Cybersecurity Strategy_ENG(uploaded)
  *   20130312_Panama_National Cybersecurity Strategy_ESP (uploaded)
  *   20240101_Togo_National Cybersecurity Strategy_FRN (uploaded)
  *   20151204_South Africa_National Cybersecurity Strategy_ENG (uploaded)
  *   20140101_Latvia_National Cybersecurity Strategy_ENG (uploaded)




# Seq Files

In [48]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf"
inaug_pdf_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("inaug_pdf_file number of unique files: ", num_unique_files)
print(inaug_pdf_filenames)

directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_txt"
processed_inaug_txt_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print("processed_inaug_txt_file number of unique files: ", num_unique_files)
print(processed_inaug_txt_filenames)

Number of files in this folder:  124
Number of unique files in this folder:  124
inaug_pdf_file number of unique files:  124
['20170101_Poland_National Cybersecurity Strategy_PLN.pdf', '20080430_Tajikistan_National Cybersecurity Strategy_RUS.pdf', '20160901_Finland_National Cybersecurity Strategy_ENG.pdf', '20160411_Colombia_National Cybersecurity Strategy_ESP.pdf', '20150101_France_National Cybersecurity Strategy_FRN.pdf', '20161101_Germany_National Cybersecurity Strategy_DEU.pdf', '20200901_Liberia_National Cybersecurity Strategy_ENG.pdf', '20190101_Estonia_National Cybersecurity Strategy_EST.pdf', '20151201_New Zealand_National Cybersecurity Strategy_ENG.pdf', '20160101_Australia_National Cybersecurity Strategy_ENG.pdf', '20200205_Brazil_National Cybersecurity Strategy_POR.pdf', '20230101_Bulgaria_National Cybersecurity Strategy_BGR.pdf', '20250101_Canada_National Cybersecurity Strategy_ENG.pdf', '20180101_Canada_National Cybersecurity Strategy_ENG.pdf', '20231204_Chile_National Cyb

In [49]:
# Double check file numbers
directory_path_pdf = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf"
directory_path_txt = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_txt"

def compare_folders(folder_a, folder_b):
    # Get file names in each folder
    files_a, files_b = os.listdir(folder_a), os.listdir(folder_b)
    files_a, files_b = set(file[:-4] for file in files_a), set(file[:-4] for file in files_b)

    print(files_a, files_b)
    # Intersection: files in both folders
    common_files = files_a & files_b

    # Files only in folder A
    only_in_a = files_a - files_b

    # Files only in folder B
    only_in_b = files_b - files_a

    return common_files, only_in_a, only_in_b

common_files, only_in_pdf, only_in_txt = compare_folders(directory_path_pdf, directory_path_txt)
print("Number of files in pdf folder: ", len(os.listdir(directory_path_pdf)))
print("Number of files in txt folder: ", len(os.listdir(directory_path_txt)))

print("Number of files in both folders: ", len(common_files))
print("Number of files only in pdf folder: ", len(only_in_pdf))
print("Number of files only in txt folder: ", len(only_in_txt))

print(only_in_pdf)
print(only_in_txt)

{'20180126_Luxembourg_National Cybersecurity Strategy_ENG', '20221001_Morocco_National Cybersecurity Strategy_FRN', '20090203_Japan_National Cybersecurity Strategy_ENG', '20160101_Malta_National Cybersecurity Strategy_ENG', '20210101_Costa Rica_National Cybersecurity Strategy_ESP', '20180101_Canada_National Cybersecurity Strategy_ENG', '20180317_New Zealand_National Cybersecurity Strategy_ENG', '20220612_Netherlands_National Cybersecurity Strategy_ENG', '20191230_Poland_National Cybersecurity Strategy_ENG', '20180501_Denmark_National Cybersecurity Strategy_ENG', '20180101_Netherlands_National Cybersecurity Strategy_ENG', '20210309_Vanuatu_National Cybersecurity Strategy_ENG', '20210901_Japan_National Cybersecurity Strategy_ENG', '20180201_Jordan_National Cybersecurity Strategy_ENG', '20141212_Russia_National Cybersecurity Strategy_RUS', '20171201_Montenegro_National Cybersecurity Strategy_ENG', '20170225_Ukraine_Information Security Doctrine_ENG', '20220103_Romania_National Cybersecuri

In [ ]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/TextFiles"
TextFiles_filenames, num_unique_files, _,_,_ = load_files_from_directory(directory_path)
print(TextFiles_filenames)

directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf"
seq_pdf_filenames, num_unique_files, _,_,_  = load_files_from_directory(directory_path)
print(seq_pdf_filenames)

text_files = set(filename[:-4] for filename in TextFiles_filenames)
seq_pdf_filenames = set(filename[:-4] for filename in seq_pdf_filenames)
intersection = list(set(text_files) & set(seq_pdf_filenames))
print('There are', len(intersection), 'files in seq pdfs that are already processed and stored in text files')
print(intersection)

to_be_processed = []
for seq_pdf_file_name in seq_pdf_filenames:
  if seq_pdf_file_name not in intersection:
    to_be_processed.append(seq_pdf_file_name + '.pdf')
print('There are', len(to_be_processed), 'files in seq pdfs that needs to be processed.')
print(to_be_processed)

# ict policy v.s. national cyber security strategies
# national cyber security strategies v.s. national cyber defense

# files that should be left in TextFiles
# Digital Agenda & Government

Number of files in this folder:  217
Number of unique files in this folder:  217
['20080101_Oman_ICT Policy_ENG.txt', '20180101_Togo_ICT Policy_ENG.txt', '20131201_Italy_ICT Policy_ENG.txt', '20140101_Qatar_ICT Policy_ENG.txt', '20160610_Japan_ICT Policy_ENG.txt', '20100511_Japan_ICT Policy_ENG.txt', '20120101_India_ICT Policy_ENG.txt', '20060202_Japan_ICT Policy_ENG.txt', '20150101_Qatar_ICT Policy_ENG.txt', '20060101_China_ICT Policy_ENG.txt', '20090203_Japan_ICT Policy_ENG.txt', '20120101_Egypt_ICT Policy_ENG.txt', '20150110_Italy_ICT Policy_ENG.txt', '20140101_Kenya_ICT Policy_ENG.txt', '20140101_Mexico_ICT Policy_ENG.txt', '20120101_Jordan_ICT Policy_ENG.txt', '20130101_Malawi_ICT Policy_ENG.txt', '20150101_Bhutan_ICT Policy_ENG.txt', '20071201_Norway_ICT Policy_ENG.txt', '20140201_Uganda_ICT Policy_ENG.txt', '20171101_Rwanda_ICT Policy_ENG.txt', '20060401_Zambia_ICT Policy_ENG.txt', '20150101_Rwanda_ICT Policy_ENG.txt', '20161101_Jordan_ICT Policy_ENG.txt', '20180101_Gambia_ICT P

In [ ]:
# load_files_from_directory('/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_txt')

Number of files in this folder:  82
Number of unique files in this folder:  82


(['20170101_Georgia_National Cybersecurity Strategy_ENG.txt',
  '20210201_Nigeria_National Cybersecurity Strategy.txt',
  '20220101_Mauritania_National Cybersecurity Strategy.txt',
  '20210731_Bangladesh_National Cybersecurity Strategy.txt',
  '20151201_New Zealand_National Cybersecurity Strategy_ENG.txt',
  '20241001_Ghana_National Cybersecurity Strategy.txt',
  '20230101_Bulgaria_National Cybersecurity Strategy_BGR.txt',
  '20221101_Gambia_National Cybersecurity Strategy.txt',
  '20240201_ROK_National Cybersecurity Strategy_KOR.txt',
  '20230720_Indonesia_National Cybersecurity Strategy_IND.txt',
  '20240101_Rwanda_National Cybersecurity Strategy.txt',
  '20160101_Australia_National Cybersecurity Strategy_ENG.txt',
  '20240101_Estonia_National Cybersecurity Strategy.txt',
  '20231101_France_National Cybersecurity Strategy.txt',
  '20231204_Chile_National Cybersecurity Strategy_ESP.txt',
  '20150101_France_National Cybersecurity Strategy_ENG.txt',
  '20230314_Latvia_National Cybersecu

In [ ]:
# load_files_from_directory('/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf')

Number of files in this folder:  82
Number of unique files in this folder:  82


(['20170101_Poland_National Cybersecurity Strategy_PLN.pdf',
  '20080430_Tajikistan_National Cybersecurity Strategy_RUS.pdf',
  '20160901_Finland_National Cybersecurity Strategy_ENG.pdf',
  '20160411_Colombia_National Cybersecurity Strategy_ESP.pdf',
  '20150101_France_National Cybersecurity Strategy_FRN.pdf',
  '20161101_Germany_National Cybersecurity Strategy_DEU.pdf',
  '20200901_Liberia_National Cybersecurity Strategy_ENG.pdf',
  '20190101_Estonia_National Cybersecurity Strategy_EST.pdf',
  '20151201_New Zealand_National Cybersecurity Strategy_ENG.pdf',
  '20160101_Australia_National Cybersecurity Strategy_ENG.pdf',
  '20210501_Belgium_National Cybersecurity Strategy.pdf',
  '20200205_Brazil_National Cybersecurity Strategy_POR.pdf',
  '20230101_Bulgaria_National Cybersecurity Strategy_BGR.pdf',
  '20250101_Canada_National Cybersecurity Strategy_ENG.pdf',
  '20180101_Canada_National Cybersecurity Strategy_ENG.pdf',
  '20231204_Chile_National Cybersecurity Strategy_ESP.pdf',
  '20230

In [ ]:
!pip install PyMuPDF

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0/20.0 MB 58.1 MB/s eta 0:00:00


In [ ]:
# import os
# folder_path = "/content/processed_sequential_pdfs"  # Replace with your desired path
# os.makedirs(folder_path, exist_ok=True)

In [ ]:
import fitz  # PyMuPDF
def process(index):
  seq_pdfs_folder_name = 'sequential_strategies_pdf'
  root = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0'
  file_path = "/".join([root, seq_pdfs_folder_name, to_be_processed[index]])
  doc = fitz.open(file_path)
  folder_path = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_txt'
  with open(folder_path + '/' + to_be_processed[index][:-4] + ".txt", "w", encoding="utf-8") as f:
      for page in doc:
          f.write(page.get_text() + '\n')

In [ ]:
for index in range(len(to_be_processed)):
  process(index)

In [ ]:
load_files_from_directory('/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_txt')

Number of files in this folder:  82
Number of unique files in this folder:  82


(['20170101_Georgia_National Cybersecurity Strategy_ENG.txt',
  '20210201_Nigeria_National Cybersecurity Strategy.txt',
  '20220101_Mauritania_National Cybersecurity Strategy.txt',
  '20210731_Bangladesh_National Cybersecurity Strategy.txt',
  '20151201_New Zealand_National Cybersecurity Strategy_ENG.txt',
  '20241001_Ghana_National Cybersecurity Strategy.txt',
  '20230101_Bulgaria_National Cybersecurity Strategy_BGR.txt',
  '20221101_Gambia_National Cybersecurity Strategy.txt',
  '20240201_ROK_National Cybersecurity Strategy_KOR.txt',
  '20230720_Indonesia_National Cybersecurity Strategy_IND.txt',
  '20240101_Rwanda_National Cybersecurity Strategy.txt',
  '20160101_Australia_National Cybersecurity Strategy_ENG.txt',
  '20240101_Estonia_National Cybersecurity Strategy.txt',
  '20231101_France_National Cybersecurity Strategy.txt',
  '20231204_Chile_National Cybersecurity Strategy_ESP.txt',
  '20150101_France_National Cybersecurity Strategy_ENG.txt',
  '20230314_Latvia_National Cybersecu

(Problem solved) Problematic Parsing:


*   **20241001_Ghana_National Cybersecurity Strategy.txt: too much graphics, parsed gibberish**
*   20240201_ROK_National Cybersecurity Strategy_KOR.txt: graphics format; less than average text
*   20231101_France_National Cybersecurity Strategy.txt: graphics format; less than average text
*   20230101_Bahrain_National Cybersecurity Strategy.txt: graphics format; less than average text
*   20221001_Morocco_National Cybersecurity Strategy_FRN.txt: shorter than average; unusual font type & size
*   **20210910_Panama_National Cybersecurity Strategy_ESP.txt: original content was scanned and formatted as pictures, not inherently in pdf format, so there is nothing to parse**
*   **20210731_Bangladesh_National Cybersecurity Strategy.txt: too much graphics and unusual format, parsed gibberish**
*   **20210201_Nigeria_National Cybersecurity Strategy.txt: non-parsable format; only the title is parsed**
*   20191230_Poland_National Cybersecurity Strategy.txt: shorter than average
*   20080430_Tajikistan_National Cybersecurity Strategy_RUS.txt: shorter than average
*   20141212_Russia_National Cybersecurity Strategy_ENG.txt: shorter than average



In [ ]:
def process_particular(particular_file_name):

  root = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf'

  file_path = "/".join([root, particular_file_name[:-4] + '.pdf'])

  print(file_path)

  doc = fitz.open(file_path)

  folder_path = '/content/problematic_parsing_pdfs'

  txt_file_path = folder_path + '/' + particular_file_name

  with open(txt_file_path, "w", encoding="utf-8") as f:
      for page in doc:
          f.write(page.get_text() + '\n')

In [ ]:
def process_particular_pdf_plumber(particular_file_name):
  root = '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf'
  file_path = "/".join([root, particular_file_name[:-4] + '.pdf'])
  print(file_path)
  folder_path = '/content/problematic_parsing_pdfs'
  txt_file_path = folder_path + '/' + particular_file_name

  with pdfplumber.open(file_path) as pdf, open(txt_file_path, "w", encoding="utf-8") as f:
      for page in pdf.pages:
          t = page.extract_text()
          if t:
              f.write(t + '\n')


In [ ]:
import fitz
particular_file_name = '20241001_Ghana_National Cybersecurity Strategy.txt'
process_particular(particular_file_name)

/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20241001_Ghana_National Cybersecurity Strategy.pdf


FileNotFoundError: no such file: '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20241001_Ghana_National Cybersecurity Strategy.pdf'

In [ ]:
import pdfplumber
particular_file_name = '20241001_Ghana_National Cybersecurity Strategy.txt'
process_particular_pdf_plumber(particular_file_name)

/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20241001_Ghana_National Cybersecurity Strategy.pdf


In [ ]:
import fitz
particular_file_name = '20210731_Bangladesh_National Cybersecurity Strategy.txt'
process_particular(particular_file_name)

/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20210731_Bangladesh_National Cybersecurity Strategy.pdf


FileNotFoundError: no such file: '/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20210731_Bangladesh_National Cybersecurity Strategy.pdf'

In [ ]:
import pdfplumber
particular_file_name = '20210731_Bangladesh_National Cybersecurity Strategy.txt'
process_particular_pdf_plumber(particular_file_name)

/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20210731_Bangladesh_National Cybersecurity Strategy.pdf


KeyboardInterrupt: 

In [ ]:
import pdfplumber
particular_file_name = '20210910_Panama_National Cybersecurity Strategy_ESP.txt'
process_particular_pdf_plumber(particular_file_name)

particular_file_name = '20210201_Nigeria_National Cybersecurity Strategy.txt'
process_particular_pdf_plumber(particular_file_name)

/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20210910_Panama_National Cybersecurity Strategy_ESP.pdf
/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/sequential_strategies_pdf/20210201_Nigeria_National Cybersecurity Strategy.pdf


# Inaug Files

In [3]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_txt"
inaug_processed_stored_txt_filenames, num_unique_files, years, countries, year_country_pairs = load_files_from_directory(directory_path)
print(inaug_processed_stored_txt_filenames)

inaug_pdf_year_country_pairs_processed_stored = year_country_pairs

Number of files in this folder:  154
Number of unique files in this folder:  154
['20170101_UAE_National Cybersecurity Strategy_ENG.txt', '20030201_USA_National Cybersecurity Strategy_ENG.txt', '20130101_India_National Cybersecurity Strategy_ENG.txt', '20140201_Kenya_National Cybersecurity Strategy_ENG.txt', '20160101_Samoa_National Cybersecurity Strategy_ENG.txt', '20170101_Egypt_National Cybersecurity Strategy_ENG.txt', '20160801_Nepal_National Cybersecurity Strategy_ENG.txt', '20170101_Chile_National Cybersecurity Strategy_ENG.txt', '20130101_Spain_National Cybersecurity Strategy_ENG.txt', '20150101_Malta_National Cybersecurity Strategy_ENG.txt', '20161227_China_National Cybersecurity Strategy_ENG.txt', '20140501_Qatar_National Cybersecurity Strategy_ENG.txt', '20131201_Italy_National Cybersecurity Strategy_ENG.txt', '20150723_Ghana_National Cybersecurity Strategy_ENG.txt', '20150301_Rwanda_National Cybersecurity Strategy_ENG.txt', '20170901_Israel_National Cybersecurity Strategy_EN

In [4]:
directory_path = "/content/drive/MyDrive/national_cybersecurity_policies_data/national_cybersecurity_strategies_v1.0/inaugural_cyber_strategies_pdf"
inaug_pdf_filenames, num_unique_files, years, countries, year_country_pairs = load_files_from_directory(directory_path)
print(len(inaug_pdf_filenames), num_unique_files)
print(inaug_pdf_filenames)
print(years)
print(countries)
print(year_country_pairs)

inaug_pdf_year_country_pairs = year_country_pairs

print(inaug_pdf_year_country_pairs)

Number of files in this folder:  150
Number of unique files in this folder:  150
150 150
['20141201_Denmark_National Cybersecurity Strategy_DAN.pdf', '20121207_Norway_National Cybersecurity Strategy_NRW.pdf', '20110802_South Korea_National Cybersecurity Strategy_ENG.pdf', '20141102_Uganda_National Cybersecurity Strategy_ENG.pdf', '20130101_Spain_National Cybersecurity Strategy_ESP.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_RUS.pdf', '20031105_Tajikistan_National Cybersecurity Strategy_ENG.pdf', '20160223_Uzbekistan_National Cybersecurity Strategy_RUS.pdf', '20170101_Chile_National Cybersecurity Strategy_ESP.pdf', '20060101_Malaysia_National Cybersecurity Strategy_ENG.pdf', '20160718_Bulgaria_National Cybersecurity Strategy_BGR.pdf', '20170417_Paraguay_National Cybersecurity Strategy_ESP.pdf', '20131201_Portugal_National Cybersecurity Strategy_POR.pdf', '20140101_Latvia_National Cybersecurity Strategy_LTV.pdf', '20180701_Macedonia_National Cybersecurity Strategy_MKD.pdf'

# Inaugural Processing (Don't Open This One)

In [ ]:
print(len(text_file_year_country_pairs), text_file_year_country_pairs)
print(len(inaug_pdf_year_country_pairs), inaug_pdf_year_country_pairs)

num_inaug_to_be_processed, num_inaug_processed = 0, 0
inaug_pdf_to_be_processed = []
for year_country_pair in inaug_pdf_year_country_pairs:
    if year_country_pair in inaug_pdf_year_country_pairs_processed_stored:
        print('Processed: ', year_country_pair)
        num_inaug_processed += 1
        continue
    if year_country_pair not in text_file_year_country_pairs:
        inaug_pdf_to_be_processed.append(year_country_pair)
        num_inaug_to_be_processed += 1
        continue
    else:
        num_inaug_processed += 1

print('Num inaug to be processed = ', num_inaug_to_be_processed)
print('Num inaug processed = ', num_inaug_processed)

print('Date & Country markers of files to be processed = ', inaug_pdf_to_be_processed)

223 ['20080101_Oman', '20180101_Togo', '20131201_Italy', '20140101_Qatar', '20160610_Japan', '20100511_Japan', '20120101_India', '20060202_Japan', '20150101_Qatar', '20060101_China', '20090203_Japan', '20120101_Egypt', '20150110_Italy', '20140101_Kenya', '20140101_Mexico', '20120101_Jordan', '20130101_Malawi', '20150101_Bhutan', '20071201_Norway', '20140201_Uganda', '20171101_Rwanda', '20060401_Zambia', '20150101_Rwanda', '20161101_Jordan', '20180101_Gambia', '20130101_Jordan', '20131014_Latvia', '20160901_Finland', '20060101_Estonia', '20170101_Namibia', '20050304_Lesotho', '20080730_Morocco', '20120101_Austria', '20151103_Belarus', '20130101_Albania', '20060109_Grenada', '20130101_Moldova', '20131201_Vanuatu', '20140801_Germany', '20150101_Liberia', '20160810_Zimbabwe', '20170101_Mongolia', '20110302_Kiribati', '20181201_Togo', '20180101_Pakistan', '20030301_Tanzania', '20160101_Malaysia', '20190101_Kiribati', '20140101_Cambodia', '20160101_Mongolia', '20160501_Cameroon', '20160501_T

In [ ]:
inaug_pdf_to_be_processed = set(inaug_pdf_to_be_processed)
inaug_pdf_to_be_processed_with_file_name = []
for inaug_pdf_filename in inaug_pdf_filenames:
  inaug_pdf_year_country = "_".join(inaug_pdf_filename.split(".")[0].split("_")[:2])
  if inaug_pdf_year_country in inaug_pdf_to_be_processed:
    inaug_pdf_to_be_processed_with_file_name.append(inaug_pdf_filename)
print("Files to be processed: ")
print(inaug_pdf_to_be_processed_with_file_name)
print("Number of files to be processed: ", len(inaug_pdf_to_be_processed_with_file_name))

Files to be processed: 
['20141201_Denmark_National Cybersecurity Strategy_DAN.pdf', '20110101_Luxembourg_National Cybersecurity Strategy_FRH.pdf', '20110802_South Korea_National Cybersecurity Strategy_ENG.pdf', '20141102_Uganda_National Cybersecurity Strategy_ENG.pdf', '20160223_Uzbekistan_National Cybersecurity Strategy_RUS.pdf', '20101220_Burkina Faso_National Cybersecurity Strategy_French.pdf', '20170417_Paraguay_National Cybersecurity Strategy_ESP.pdf', '20131201_Portugal_National Cybersecurity Strategy_POR.pdf', '20130101_Turkey_National Cybersecurity Strategy_TUR.pdf', '20130101_Turkey_National Cybersecurity Strategy_ENG.pdf', '20190101_Mauritania_National Cybersecurity Strategy_FRN.pdf', '20140101_Syria_National Cybersecurity Strategy_Arabic.pdf', '20140417_Uruguay_National Cybersecurity Strategy_ESP.pdf', '20170225_Ukraine_Information Security Doctrine_ENG.pdf', '20130101_Russia_National Cybersecurity Strategy_ENG.pdf', '20110101_France_National Cybersecurity Strategy_FRN.pdf'

In [ ]:
print(inaug_pdf_year_country_pairs_processed_stored)

['20030201_USA', '20120423_Cyprus', '20000909_Russia']




*   TextFile Folder size = 223
*   Seq all pdf folder size = 82 (71 to be processed, 11 already processed, 71 + 11)
*   Inaug all pdf folder size = 195 (71 to be processed, 124 already processed, 71 + 124)

* Total of pdfs from both types to be processed = 71 + 71 = 142

*   Number of seq pdfs already processed = 11
*   Number of inaug pdfs already processed = 124 (after removing duplicates, it can be 123)
*   Total number of pdfs already processed = 11 + 124 = 135 (or 134)


In [ ]:
num_of_files_in_textfile_butnotin_seq_or_inaug = 0
num_seq_processed_0 = 0
num_inaug_processed_0 = 0
num_in_both_0 = 0 # this should be 0
print(len(text_file_year_country_pairs))
print(len(seq_pdf_year_country_pairs))
print(len(inaug_pdf_year_country_pairs))


for text_file_year_country_pair in text_file_year_country_pairs:
  if text_file_year_country_pair in seq_pdf_year_country_pairs and text_file_year_country_pair in inaug_pdf_year_country_pairs:
    num_in_both_0 += 1
    continue
  if text_file_year_country_pair in seq_pdf_year_country_pairs:
    num_seq_processed_0 += 1
  if text_file_year_country_pair in inaug_pdf_year_country_pairs:
    num_inaug_processed_0 += 1

  if text_file_year_country_pair not in inaug_pdf_year_country_pairs and text_file_year_country_pair not in seq_pdf_year_country_pairs:
    num_of_files_in_textfile_butnotin_seq_or_inaug += 1

print()
print(num_of_files_in_textfile_butnotin_seq_or_inaug)
print(num_seq_processed_0)
print(num_inaug_processed_0)
print(num_in_both_0)

print(num_of_files_in_textfile_butnotin_seq_or_inaug + num_seq_processed_0 + num_inaug_processed_0 + num_in_both_0)

223
82
195

102
8
110
3
223
